Workflow/Sources/WorkflowObserver.swift

```diff
@@ -0,0 +1,367 @@
+    ///   - session: The `WorkflowSession` corresponding to the backing `WorkflowNode`.
+    func workflowDidReceiveAction<Action: WorkflowAction>(
+        _ action: Action,
+        workflow: Action.WorkflowType,
+        session: WorkflowSession
+    )
+
+    /// Called when a `WorkflowAction` will be applied to its corresponding Workflow's State
+    /// - Parameters:
+    ///   - action: The action that will be applied.
+    ///   - workflow: The action's corresponding `Workflow`.
+    ///   - state: The state to which the action will be applied.
+    ///   - session: The `WorkflowSession` corresponding to the backing `WorkflowNode`.
+    /// - Returns: An optional closure to be called immediately after the action is applied to the `State`, and an optional `Output` has been produced. The closure takes the updated state and optional output as its arguments.
+    func workflowWillApplyAction<Action: WorkflowAction>(
+        _ action: Action,
+        workflow: Action.WorkflowType,
+        state: Action.WorkflowType.State,
+        session: WorkflowSession
+    ) -> ((Action.WorkflowType.State, Action.WorkflowType.Output?) -> Void)?
+}
+
+// MARK: - WorkflowSession
+
+/// A `WorkflowSession`encapsulates the information that gives a `WorkflowNode` its identity.
+/// In particular, it captures:
+/// - The type of the corresponding `Workflow` conformance.
+/// - The `String` key used when the workflow was rendered.
+/// - An `Identifier` type that is unique across program execution.
+/// - An optional reference to a parent `WorkflowSession`, to differentiate root nodes.
+public struct WorkflowSession {
+    public struct Identifier: Hashable {
+        private static var _nextRawID: UInt64 = 0
+        private static func _makeNextSessionID() -> UInt64 {
+            let nextID = _nextRawID
+            _nextRawID += 1
+            return nextID
+        }
+
+        let rawIdentifier: UInt64 = Self._makeNextSessionID()
+    }
+
+    /// As structs cannot contain stored properties of their own type, we use an indirect enum
+    /// to allow referencing the parent session.
+    private indirect enum IndirectParent {
+        case some(WorkflowSession)
+        case none
+
+        init(_ parent: WorkflowSession?) {
+            switch parent {
+            case .some(let value):
+                self = .some(value)
+            case .none:
+                self = .none
+            }
+        }
+    }
+
+    public let workflowType: Any.Type
+
+    public let renderKey: String
+
+    public let sessionID = Identifier()
+
+    private let _indirectParent: IndirectParent
+    public var parent: WorkflowSession? {
+        switch _indirectParent {
+        case .some(let parent):
+            return parent
+        case .none:
+            return nil
+        }
+    }
+
+    /// Creates a new `WorkflowSession` instance. Note, construction of this type
+    /// is not safe to perform concurrently with respect to other instance initialization.
```

Comment:
Why is that? Should we be doing something to guard against misusage?

Reply:
internally it uses a global variable to generate the unique ids which has no synchronization. i suppose we could add something more, though i'm not sure what in particular would be of value. in general i think of Workflow as a main-thread-only API at the moment (the runtime enforces this in various places), so IMO somehow instantiating this off the main thread would already be somewhat unexpected.